src/Orchard.Web/Modules/Orchard.Autoroute/Handlers/AutoroutePartHandler.cs

```diff
@@ -44,69 +53,109 @@ public class AutoroutePartHandler : ContentHandler {
             OnUnpublished<AutoroutePart>((ctx, part) => RemoveAlias(part));
 
             // Register alias as identity
-            OnGetContentItemMetadata<AutoroutePart>((ctx, part) => {
+            OnGetContentItemMetadata<AutoroutePart>((ctx, part) =>
+            {
                 if (part.DisplayAlias != null)
                     ctx.Metadata.Identity.Add("alias", part.DisplayAlias);
             });
         }
 
-        private void CreateAlias(AutoroutePart part) {
+        private void CreateAlias(AutoroutePart part)
+        {
             ProcessAlias(part);
         }
 
-        private void PublishAlias(AutoroutePart part) {
-            ProcessAlias(part);
+        private string LockString {
+            get {
+                return string.Join(".",
+                    _orchardServices.WorkContext?.CurrentSite?.BaseUrl ?? "",
+                    _orchardServices.WorkContext?.CurrentSite?.SiteName ?? "",
+                    "AutoroutePartHandler");
+            }
+        }
 
-            // Should it become the home page?
-            if (part.PromoteToHomePage) {
-                // Get the current homepage an unmark it as the homepage.
-                var currentHomePage = _homeAliasService.GetHomePage(VersionOptions.Latest);
-                if(currentHomePage != null && currentHomePage.Id != part.Id) {
-                    var autoroutePart = currentHomePage.As<AutoroutePart>();
-
-                    if (autoroutePart != null) {
-                        autoroutePart.PromoteToHomePage = false;
-                        if(autoroutePart.IsPublished())
-                            _orchardServices.ContentManager.Publish(autoroutePart.ContentItem);
+        private void PublishAlias(AutoroutePart part)
+        {
+            // Lock this portion of the code to prevent race conditions where we are reading the aliases to compute
+            // stuff for a content, while another content is publishing. 
+            // If the generated alias is empty, compute a new one.
+            _lockingProvider.Lock(LockString,
+                () =>
+                {
+                    ProcessAlias(part);
+
+                    // Should it become the home page?
+                    if (part.PromoteToHomePage)
+                    {
+                        // Get the current homepage an unmark it as the homepage.
+                        var currentHomePageId = _homeAliasService.GetHomePageId();
+                        if (currentHomePageId != part.Id)
+                        {
+                            var autoroutePart = part.As<AutoroutePart>();
```

Comment:
We still need to fix this, because `autoroutePart` here should be the pre-existing home page, but the code as it is in this PR sets it to the same part that we are trying to update and set as homepage.